designs/2024-config-lookup-from-file/README.md

```diff
@@ -0,0 +1,134 @@
+
+In most of the `ESLint` class logic, `FlatConfigArray#getConfig()` will need to be replaced by `ConfigLoader#loadConfigArrayForPath()` to ensure that the file system is always searched to find the correct configuration.
+
+It's necessary to return a `FlatConfigArray` because we [pass a `FlatConfigArray` to `Linter`](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L498) and also use it when we [filter out code blocks](https://github.com/eslint/eslint/blob/455f7fd1662069e9e0f4dc912ecda72962679fbe/lib/eslint/eslint.js#L511-L513). Because `Linter` is a synchronous API, we need to maintain the synchronous calls, and the easiest way to do that is to access the `FlatConfigArray` directly.
+
+### Core Changes
+
+In order to make all of this work, we'll need to make the following changes in the core:
+
+1. Create a new `lib/config/config-loader.js` file to contain both `ConfigLoader` and `LegacyConfigLoader`.
+1. Move `findFlatConfigFile()`, `loadFlatConfigFile()`, `locateConfigFileToUse()`, and `calculateConfigArray()` from `lib/eslint/eslint.js` to `lib/config/config-loader.js`.
+1. Create a new feature flag called `unstable_config_lookup_from_file` (once https://github.com/eslint/eslint/pull/18516 is merged).
+1. Update `lib/eslint/eslint.js`:
+    1. Create a new `ConfigLoader` based on the presence or absence of the flag. This should be used in the `ESLint#lintText()` and `ESLint#lintFiles()` methods.
+    1. Update `ESLint#calculateConfigForFile()` to use the config loader.
+    1. Update `ESLint#findConfigFile()` to use the config loader. To preserve backwards compatibility, when called without an argument, this method will start the config file search from the current working directory; it will also accept an argument, which is the path to search from.
+    1. Update `getOrFindUsedDeprecatedRules()` to use the config loader.
+1. Update `findFiles()` in `lib/eslint/eslint-helpers.js` to use the config loader. This also requires a change to the file system walking logic because `fswalk` filter functions are synchronous, but we'll need them to be asynchronous to use with the config loader. I plan on using [`humanfs`](https://github.com/humanwhocodes/humanfs/blob/main/docs/directory-operations.md#reading-directory-entries-recursively) (which I wrote).
```

Comment:
If there's no config file in the root, but there is a config file in a subdirectory, e.g., `src/eslint.config.js`, what happens when user runs:

1. `eslint .` from the project root. Does it error saying that there's no config file?
2. `eslint src` from the project root. Does it load `src/eslint.config.js` and lint files in `src`?

Reply:
This works the same way as in eslintrc: A config file is looked up from the file being linted not from the patterns passed on the command line. So `eslint .` and `eslint src` work the same for the same files. If there's `./src/foo.js` and `./src/eslint.config.js`, then when ESLint goes to lint `./src/foo.js` it loads `./src/eslint.config.js`. Then, in the same run, if there's a `./tests/eslint.config.js`, that will be used for `./tests/foo.test.js`.

Reply:
In that case, how does the search determine which directories should be traversed when running `eslint .` and there is no config file in the root? Will it skip those matched by [default ignores](https://github.com/eslint/eslint/blob/375227f94da3c1c4ff6c61a29b272889fa48ca79/lib/config/default-config.js#L56-L62)?

Reply:
Everything is dependent on the directory that's visited. For each directory, we search up to find a config file and use that to determine whether or not the directory should be skipped. The same is true for files. We always search for the relevant config file.

Reply:
That would mean we'd search `node_modules` too, and lint files in it if some packages have an `eslint.config.js` file published?

Reply:
No. There are two scenarios for when we traverse into `node_modules`.

1. A config file is found in its ancestry. Because `node_modules` is a default ignore, we won't continue into the subdirectories.
2. A config file is not found so an error is thrown.

Reply:
I think we can better answer these questions with a prototype, so I'm going to work on that.

Reply:
I've tried the above scenario with the `issue18385b` branch. Both  `npx eslint . --flag unstable_config_lookup_from_file` and `npx eslint src --flag unstable_config_lookup_from_file` result in `ESLint couldn't find an eslint.config.(js|mjs|cjs) file.` error.

Furthermore, if I add an empty (`export default []`) config file in the root, both commands now work but neither seems to use `src/eslint.config.js`.

Reply:
I've tried the latest version of the `issue18385b` branch with a test project that has a `src/eslint.config.js` file but no root `eslint.config.js` file.

Running `eslint .` from the root fails this way:

```
$ npx eslint . --flag unstable_config_lookup_from_file --debug
  eslint:cli CLI args: [ '.', '--flag', 'unstable_config_lookup_from_file', '--debug' ] +0ms
  eslint:cli Using flat config? true +3ms
  eslint:cli Running on files +4ms
  eslint:cli Checking for inactive flags +1ms
  eslint:eslint Using file patterns: . +0ms
  eslint:eslint Deleting cache file at C:\projects\tmp\tmp\.eslintcache +1ms
  eslint:config-loader Calculating config for directory C:/projects/tmp/tmp/node_modules +0ms
  eslint:config-loader Searching for eslint.config.js +1ms

Oops! Something went wrong! :(

ESLint: 9.5.0

ESLint couldn't find an eslint.config.(js|mjs|cjs) file.

From ESLint v9.0.0, the default configuration file is now eslint.config.js.
If you are using a .eslintrc.* file, please follow the migration guide
to update your configuration file to the new format:

https://eslint.org/docs/latest/use/configure/migration-guide

If you still have problems after following the migration guide, please stop by
https://eslint.org/chat/help to chat with the team.
```

Does this mean that there must be an `eslint.config.js` in the root (or ancestry) of the search pattern?

Running `eslint src` from the root fails in the same way:

```
$ npx eslint src --flag unstable_config_lookup_from_file --debug
  eslint:cli CLI args: [ 'src', '--flag', 'unstable_config_lookup_from_file', '--debug' ] +0ms
  eslint:cli Using flat config? true +3ms
  eslint:cli Running on files +3ms
  eslint:cli Checking for inactive flags +1ms
  eslint:eslint Using file patterns: src +0ms
  eslint:eslint Deleting cache file at C:\projects\tmp\tmp\.eslintcache +1ms
  eslint:config-loader Calculating config for directory C:/projects/tmp/tmp/node_modules +0ms
  eslint:config-loader Searching for eslint.config.js +0ms

Oops! Something went wrong! :(

ESLint: 9.5.0

ESLint couldn't find an eslint.config.(js|mjs|cjs) file.

From ESLint v9.0.0, the default configuration file is now eslint.config.js.
If you are using a .eslintrc.* file, please follow the migration guide
to update your configuration file to the new format:

https://eslint.org/docs/latest/use/configure/migration-guide

If you still have problems after following the migration guide, please stop by
https://eslint.org/chat/help to chat with the team.
```

This is, I believe, an unexpected behavior because the argument is `src` so the absence of config files in other directories shouldn't matter I think. Running `cd src && npx eslint . --flag unstable_config_lookup_from_file --debug` works well.

Reply:
Okay, so I've updated the behavior such that traversing directories does not require a config file but will honor one if it's found. There will be no error thrown when a config file is not found when traversing directories.

There will be an error thrown when attempting to lint a file that doesn't have a config file in its ancestry. 

I believe this behavior mimics what we have now.